### ISSUE TITLE
Bug inserting a item-index in a string-list

### ISSUE THREAD

**Issue by fredvs** [mse-org/mseide-msegui#27]
About the break-order when inserting a item.
You propose to compare with the msei18n_i18n.trd.

Ok, but see how it is:

```
"mainfo,mainmenu1,menu.submenu.items,0:,caption",vaString,F,,&Open,&Open
"mainfo,mainmenu1,menu.submenu.items,0:,name",vaString,T,,open,
"mainfo,mainmenu1,menu.submenu.items,0:,state",vaSet,F,,"[as_localcaption,as_localonexecute]",
"mainfo,mainmenu1,menu.submenu.items,0:,onexecute",vaIdent,F,,onprojectopen,
"mainfo,mainmenu1,menu.submenu.items,1:",vaList,F,,,
"mainfo,mainmenu1,menu.submenu.items,1:,caption",vaString,F,,&Save,&Save
"mainfo,mainmenu1,menu.submenu.items,1:,name",vaString,T,,save,
"mainfo,mainmenu1,menu.submenu.items,1:,state",vaSet,F,,"[as_localcaption,as_localonexecute]",
"mainfo,mainmenu1,menu.submenu.items,1:,onexecute",vaIdent,F,,onprojectsave,

```

For each submenu.items,0, there is 3 lines. If one item is inserted, it will kill the order.
I do not see how to do to compare, first with `name`, then `caption` ?

**Comment by glixx** [comment]
Give for all translated strings default Value. And how you did it for restoration NT flag from trd file do same job for translated strings.

**Comment by fredvs** [comment]
> And how you did it for restoration NT flag from trd file do same job for translated strings.

This was a totally different problem.
If NT flag is enabled in a list, it breaks the order of the index but not the value.

Here, when inserting a new item it breaks the order and add a new value that detune all the other values.

Note that if you have set NT to false in the item list, all will be on wrong place too.
My NT fixes works only if you does not insert a item.

All idea are welcome.

**Comment by glixx** [comment]
Try this way https://github.com/mse-org/mseide-msegui/issues/27#issuecomment-474636482
I have no better idea. Just do not read file again but do this job while you read NT from trd file.

**Comment by fredvs** [comment]
I may try but it will need lot of resource, each item must be compared with the list in trd.
Maybe a button "Reorder the list".

**Comment by glixx** [comment]
> Maybe a button "Reorder the list".

Absolutely no.

**Comment by glixx** [comment]
Only trd file keeps translated strings and needs to use them.

**Comment by fredvs** [comment]
OK, fixed in commit a138852

**Comment by glixx** [comment]
Encoding problems and some strings lost translation.
![Screenshot_20190320_224730](https://user-images.githubusercontent.com/2681872/54724346-cb934e00-4b7b-11e9-9656-a766e2281c31.png)
NT flags were not restored.
I just open msei18n project, 2013

**Comment by fredvs** [comment]
> NT flags were not restored.
> I just open msei18n project, 2013

It is what I wrote there : https://github.com/mse-org/mseide-msegui/commit/a138852afe381f42cefc8630d9e25d1b72c88b44#commitcomment-32849571

Better is not possible, all what could have conflict is set as default.

**Comment by glixx** [comment]
Who needs order if big mess?

**Comment by fredvs** [comment]
> Encoding problems

Can you describe?

> Who needs order if big mess?

I think you do not have understand the problems.

**Comment by glixx** [comment]
You see ???? in picture.
For example "E&xit" was translated. Why it has conflict?

**Comment by fredvs** [comment]
From your Russian file:

![msei18nru](https://user-images.githubusercontent.com/3421249/54724925-66386080-4b6d-11e9-83cf-9fe56dcd47e6.png)

**Comment by glixx** [comment]
My picture is https://github.com/mse-org/mseide-msegui/issues/27#issuecomment-475060030

**Comment by fredvs** [comment]
> "E&xit" was translated. Why it has conflict?

It has not the same index that existed before (there was a number index -1 before).
Is it so terrible to check only one or two rows, there are red, it is easy.

**Comment by fredvs** [comment]
> My picture is #27 (comment)

Aaargh, maybe a utf bad transform, strange that it works here.

**Comment by glixx** [comment]
Reading trd file by index is wrong. Needs to read by component name and Value.

**Comment by fredvs** [comment]
> Reading trd file by index is wrong. Needs to read by component name and Value.

It is what it does now!

**Comment by fredvs** [comment]
Why dont you look at code?

**Comment by glixx** [comment]
You should not use index for restoration translations. Just component name and Value.
I support 300+ programs, I have no time.

**Comment by fredvs** [comment]
> You should not use index for restoration translations. Just component name and Value.

это то, что сделано в последнем коммите

**Comment by glixx** [comment]
"E&xit" has no conflicts. Do not use index.

**Comment by fredvs** [comment]
> Just component name

In a string-list there are no component name, only a number, the index

**Comment by fredvs** [comment]
> "E&xit" has no conflicts. Do not use index.

OK, I will do but I do not use the index.

**Comment by fredvs** [comment]
> You see ???? on picture.

Is it better with last commit?

**Comment by glixx** [comment]
> Is it better with last commit?

Same

**Comment by fredvs** [comment]
And with this last commit from now?

**Comment by glixx** [comment]
> And with this last commit from now?

Same

**Comment by fredvs** [comment]
> Same

How can I do to have the same issue ?
With your Russian file it works perfectly here.

It is very difficult to fix something that not be reproduce.

**Comment by glixx** [comment]
I don't know why it's ok on your side, but not on my side.

**Comment by fredvs** [comment]
And with last commit from now?

**Comment by fredvs** [comment]
It is very boring because if copy() does not work ok, how to fix the bugs.

**Comment by glixx** [comment]
> And with last commit from now?

Sorry, but same.

**Comment by fredvs** [comment]
;(

I will stop for now, I do no understand why it is ok here.
Is it for all your projects, for all Russian strings, are there some ok?

**Comment by fredvs** [comment]
If you edit a row `ru` and change with russian character, is it ok when save + reload?

**Comment by glixx** [comment]
Yes, for all projects.
Editing is ok.

**Comment by fredvs** [comment]
> Editing is ok.

And save + reload?

**Comment by glixx** [comment]
It save in wrong encoding. So reload is wrong too.

**Comment by fredvs** [comment]
> It saves in wrong encoding. So reload is wrong too.

Did you check the trd file, is it saved ok there?

**Comment by glixx** [comment]
trd file has ????? too.

**Comment by fredvs** [comment]
> trd file has ????? too.

Did you try with a trd with russian characters?

**Comment by fredvs** [comment]
> trd file has ????? too

Also for the row that you edited and replaced by russian characters?

**Comment by fredvs** [comment]
With this file it works ok here:

[lang_ru_msei18n.zip](https://github.com/mse-org/mseide-msegui/files/2990611/lang_ru_msei18n.zip)

**Comment by glixx** [comment]
![Screenshot_20190321_014029](https://user-images.githubusercontent.com/2681872/54729247-7b27ea80-4b93-11e9-9b66-50fd73217d64.png)

**Comment by fredvs** [comment]
![msei18nru2](https://user-images.githubusercontent.com/3421249/54729463-f2558280-4b83-11e9-9c15-249ee9ebe9c2.png)

**Comment by fredvs** [comment]
Why?

**Comment by glixx** [comment]
Try LC_ALL=C msei18n

**Comment by glixx** [comment]
I found reason. I deleted libi18n_ru.so and encoding is ok.

**Comment by fredvs** [comment]
What, is it ok now?

**Comment by fredvs** [comment]
> Try LC_ALL=C msei18n

What is that?

**Comment by glixx** [comment]
No. Not ok. If libi18n_ru.so is present, then encoding is wrong.

> What is that?

Run msei18n in English default mode.

**Comment by fredvs** [comment]
> No. Not ok. If libi18n_ru.so is present, then encoding is wrong.
> 
> > What is that?
> 
> Run msei18n in English default mode.

Sorry I am totally lost.
I run in default Englsh mode, without loading any libi18n_ru.so

**Comment by glixx** [comment]
But you don't have libi18n_ru.so. If you will put libi18n_fr.so I guess you'll reproduce problem.

**Comment by fredvs** [comment]
OK, I did this, compile your file to get `libi18n_ru.so` and load `msei18n` forcing Russian language:

```
program msei18n;
...
 Gettext.GetLanguageIDs(MSELang,MSEFallbackLang);
 MSEFallbackLang := 'ru';
 If loadlangunit('i18n_'+MSEFallbackLang,true) then
                                                setlangconsts(MSEFallbackLang);
...
```

Result:

![msei18nru3](https://user-images.githubusercontent.com/3421249/54730073-1070b200-4b87-11e9-8126-3a6c4e5be92e.png)

**Comment by fredvs** [comment]
> try LC_ALL=C msei18n now

I do not understand, what is LC_ALL=C ?

**Comment by glixx** [comment]
To set C locale forced and environment variables.
Did you reproduce?

**Comment by fredvs** [comment]
> To set C locale forced and environment variables.
> Did you reproduce?

I do not understand how to set C locale forced and environment variables.

How do you do that?

**Comment by glixx** [comment]
Type in console:
`LC_ALL=C msei18n`

**Comment by fredvs** [comment]
Yes, indeed, lot of ?????
But what is use of this LC_ALL=C?

Please, what is the command to go back?

**Comment by glixx** [comment]
I use it when I need to run application without localization.

**Comment by fredvs** [comment]
OK, now I get somethng to try, I have the same problem if I do LC_ALL=C msei18n without any libi18n_ru.so installed.

So I can test.

Huh, are you really sure that it worked before?

**Comment by glixx** [comment]
Yes. I used it always and yesterday it worked.

**Comment by fredvs** [comment]
OK, I will try tomorrow.
Thanks for your patience.

**Comment by glixx** [comment]
For lists:
You should read Value from trd file and for all items of list to search same Value, then read next Value from trd file ...

**Comment by fredvs** [comment]
> For lists:
You should read Value from trd file and for all items of list to search same Value, then read next Value 

It is what I do and did explained lot of time now.
Maybe you could read the code.
There s a bug with the mse-nodes.
I will do a break with mse, sorry.

**Comment by fredvs** [comment]
> Type in console:
> `LC_ALL=C msei18n`

It should be fixed in last release c498818

**Comment by fredvs** [comment]
> For lists:
> You should read Value from trd file and for all items of list to search same Value, then read next Value from trd file ...

OK, I will try to explain one more time.
The last commits do search by same original Value.
Now, even if you insert a item at beginning of list, all the order and good translation is preserved.
It was not the case of previous algorithm.

When a new item is inserted in a list, the default Value will be used and the row will be red.
But, because a bug in node-assign that I do not have fixed yet, it s not possible to do a search for items that does not exist in the `trd` file.
And that occurs for all new items that are added in list, in fact it is not inserted by append at end.

So, at the moment, for each item added, the last one in the list, even if t was already translated, can not be updated and the default value will be used as translated value.

It is the case of `E&xit` in your `trd` file, that is the last row of a list where a item was inserted here "Refresh".

I agree it is not perfect but now it is no more a mess if you insert something.
And yes, it is boring to have to retranslate the last row because it was already done.

And working on the grid is not the solution, it must be on the node (and it is why it is much more complicated).

**Comment by fredvs** [comment]
msei18n works like that

```
 while not trd.eof do begin
...
if node.findsubnode(aname_from_trd)
  then change_value_of_node_with_value_of_trd;
```

It should be like this

```
 while node.item <  node.itemcount do begin
...
if find_in_trd(anode_value)
  then change_value_of_node_with_value_of_trd;
```

**Comment by glixx** [comment]
Loosing translated strings is bug. It's good if you know how to fix it.

**Comment by fredvs** [comment]
> Loosing translated strings is bug. It's good if you know how to fix it.

I know perfecty the problem, all is ready to fix if but  am blocked by the tree/node of mse.
It is extremely difficult to understand hoe it works and I did not find any example.

I want to know how to do with mse to access each sub-item of each sub-node:

 while node.item <  node.itemcount do begin
...
if find_in_trd(node.item.value)
  then change_value_of_node_with_value_of_trd;

**Comment by glixx** [comment]
Ask in mailing list if somebody knows.

**Comment by fredvs** [comment]
OK, done.

Apart for the last item of a list, do you see other problems, the bad-character is it fixed?

**Comment by glixx** [comment]
I opened Russian the latest mseide project. It lost translation not only for last item of list.

**Comment by fredvs** [comment]
I just commit with hope that it fix missing items, this working on the node.

Is it ok for for Russian value `?????` for you?

**Comment by fredvs** [comment]
Oooops, no, no good, lease wait for next commit.

**Comment by fredvs** [comment]
Re-oops, no it was ok, the last commit should be ok.

**Comment by glixx** [comment]
I don't see differencies. It loses any translations.

> Is it ok for for Russian value ????? for you?

Yes.

**Comment by glixx** [comment]
![Screenshot_20190322_014709](https://user-images.githubusercontent.com/2681872/54795513-92cba580-4c5d-11e9-9f78-6436407dddf0.png)

**Comment by fredvs** [comment]
OK, I did try with you tpd for mse ide.
I think I see why some are loosed.

In the tpd, some original value has " " and the node.name do not add it, for example:

`"templateeditorfo,c,strings.data,0:",vaString,F,,"Do you want to delete ""","Вы хотите удалить """`

All other do not have " " for value:

"templateeditorfo,c,strings.data,1:",vaString,F,,Code Template Editor,Редактор шаблонов кода

So, I hope the solution is to ignore those " " for the search.

You now what, I think that we will get it.

**Comment by glixx** [comment]
Yes. Some strings have begin " and end " if string has " inside.

**Comment by fredvs** [comment]
See for example in your picture, the first red line.
The default value for the node is  `Recursive form hierarchy for "  with 1 `"`
`
But in your trd, it is "Recursive form hierarchy for """ with 4 "

"mainfo,c,strings.data,6:",vaString,F,,"Recursive form hierarchy for ""","Рекурсия в иерархии форм для """

So it is normal that it does not find it because it is not exactly the same.
Ok, asap, I will do a other search ignoring the ""

**Comment by glixx** [comment]
No. They are same. It just format of trd file.

**Comment by fredvs** [comment]
> No. They are same. It just format of trd file.

Not in the algorithm, I just tested, the search must be done, the "" should be ignored to compare value from the node with value of the trd.

I will not change anything on the trd, do not afraid, only the search filter.

**Comment by fredvs** [comment]
> I don't see differencies. It loses any translations.

Yes maybe with the `trp` of `mseide`.
But if you try  the `trp` of `msei18n`, you will note that all the last items of list are translated now, even` E&xit.`

**Comment by glixx** [comment]
> But if you try the trp of msei18n, you will note that all the last items of list are translated now, even E&xit.

No. Same. Tested 2013 version.

**Comment by fredvs** [comment]
> > But if you try the trp of msei18n, you will note that all the last items of list are translated now, even E&xit.
> 
> No. Same. Tested 2013 version.

Please, close the project and reload it.

**Comment by fredvs** [comment]
Re-ooops,  I see, sorry, write you later.
(What a big fight)

**Comment by glixx** [comment]
Yes. On the second start I see changes.

**Comment by fredvs** [comment]
> nd start I see changes.

Yes but it should be aways, I know why.

**Comment by fredvs** [comment]
OK, it should works always with last commit.

**Comment by glixx** [comment]
Let me know about fixed filter for "

**Comment by fredvs** [comment]
Tomorrow I will attack the `""` problem (but ignoring is not the solution because sometime it is needed).

**Comment by glixx** [comment]
It should read "File """ as File "

**Comment by fredvs** [comment]
> It should read "File """ as File "

Exactly!

**Comment by glixx** [comment]
And there is problem of reading multiline strings, cause they are inside ""

**Comment by fredvs** [comment]
> And there is problem of reading multiline strings, cause they are inside ""

Yes and it s a other battle.

String single-line with `"` is already solved. 

But for multi-line value it is more complicated because the value is stored on the `trp`  with sometime `#10`  character, or `#10#13`  or `#13` ...

Needs (lot of) cups coffee.

**Comment by fredvs** [comment]
Good coffee!

![msei18nru4](https://user-images.githubusercontent.com/3421249/54842819-47ec7500-4cd3-11e9-8d4c-804196bfa6b5.png)

**Comment by fredvs** [comment]
Note that now the search is by the default-value not the index.

So for example in your `msei18nru.trd`, one of the default value is `Show Color` but in the `mfm` file, the default value is `show color`  ( in lowercase ).

In that case, `Show Color`  will not be found, you may fix it changing to `show color` in `trd` file.

**Comment by glixx** [comment]
Nice to see picture with potentially fixes. I will test on Monday.
"Show Color" must not be found. This string was changed.

**Comment by glixx** [comment]
Many strings are still problematic:
![Screenshot_20190324_231524](https://user-images.githubusercontent.com/2681872/54887443-3ad2b000-4ea4-11e9-957d-60b97c85b16a.png)
![Screenshot_20190324_231629](https://user-images.githubusercontent.com/2681872/54887446-3c9c7380-4ea4-11e9-8e51-cfadc5e99802.png)

**Comment by glixx** [comment]
msei18n project is better, than mseide. mseide has many strings that were read from TRD.

**Comment by fredvs** [comment]
> Many strings are still problematic:
> ![Screenshot_20190324_231524](https://user-images.githubusercontent.com/2681872/54887443-3ad2b000-4ea4-11e9-957d-60b97c85b16a.png)
> ![Screenshot_20190324_231629](https://user-images.githubusercontent.com/2681872/54887446-3c9c7380-4ea4-11e9-8e51-cfadc5e99802.png)

Should be fixed in last commits.

**Comment by glixx** [comment]
Still 12 strings are problematic:
![Screenshot_20190326_000752](https://user-images.githubusercontent.com/2681872/54962284-4215be80-4f75-11e9-97ce-7f1c06958fc7.png)

**Comment by glixx** [comment]
For projects of 2013 everything is very bad. It lost all translations, even import does not work.

**Comment by fredvs** [comment]
> even import does not work

?

Maybe your file is corrupted.
Here importing your po file:

![msei18nru6](https://user-images.githubusercontent.com/3421249/54964451-d62f5800-4f6c-11e9-8a1a-cd2d6d86ecec.png)

**Comment by glixx** [comment]
I'm not about import from PO. It has no translations on start and import from csv without effects.

**Comment by fredvs** [comment]
Ha, ok, I see, after import, all seems ok, but even after save, reloading the project, translations are missing.

Is it that?

O, I have to check this.

**Comment by glixx** [comment]
I have no translations even on the first start.

**Comment by fredvs** [comment]
Please sent me your problematic file.
With the one you give me last month, it works.

**Comment by glixx** [comment]
Main bug was fixed. Other problems are in other issues.
